text.c

```diff
@@ -557,38 +560,102 @@ bool text_insert(Text *txt, size_t pos, const char *data, size_t len) {
 	return true;
 }
 
-size_t text_undo(Text *txt) {
+size_t action_undo(Text *txt, Action *a) {
 	size_t pos = EPOS;
-	/* taking a snapshot makes sure that txt->current_action is reset */
-	text_snapshot(txt);
-	Action *a = action_pop(&txt->undo);
-	if (!a)
-		return pos;
 	for (Change *c = a->change; c; c = c->next) {
 		span_swap(txt, &c->new, &c->old);
 		pos = c->pos;
 	}
+	return pos;
+}
+
+size_t action_redo(Text *txt, Action *a) {
+	size_t pos = EPOS;
+	Change *c = a->change;
+	while (c->next)
+		c = c->next;
+	for ( ; c; c = c->prev) {
+		span_swap(txt, &c->old, &c->new);
+		pos = c->pos;
+	}
+	return pos;
+}
 
-	action_push(&txt->redo, a);
+size_t text_undo(Text *txt) {
+	size_t pos = EPOS;
+	/* taking a snapshot makes sure that txt->current_action is reset */
+	text_snapshot(txt);
+	Action *a = txt->history->prev;
+	if (!a)
+		return pos;
+	pos = action_undo(txt, txt->history);
+	txt->history = a;
 	lineno_cache_invalidate(&txt->lines);
 	return pos;
 }
 
 size_t text_redo(Text *txt) {
 	size_t pos = EPOS;
-	Action *a = action_pop(&txt->redo);
+	/* taking a snapshot makes sure that txt->current_action is reset */
+	text_snapshot(txt);
+	Action *a = txt->history->next;
 	if (!a)
 		return pos;
-	for (Change *c = a->change; c; c = c->next) {
-		span_swap(txt, &c->old, &c->new);
-		pos = c->pos;
+	pos = action_redo(txt, a);
+	txt->history = a;
+	lineno_cache_invalidate(&txt->lines);
+	return pos;
+}
+
+bool history_change_branch(Action *a) {
+	bool changed = false;
+	while (a->prev) {
+		if (a->prev->next != a)
```

Comment:
You want to enclose the if block with curly braces, otherwise changed will always be true.